### PR TITLE
Create circl_passive_ssl.py

### PR DESCRIPTION
**Pull request by threathive**
Passive SSL plugin using Circl.lu api. I would create theses into certificate objects , but the api works on sha1 hashes and not actual data except in the case of the fetch certificate def. In any case to turn the output into a cert object would be a bit more work so for now this will just save the cert details in text nodes and in the raw data.

### CODE REVIEW

**Comment by p-l-** [comment]
FWIW a `CertificateSubject` observable type exists and can be used to store both a certificate's subject and the issuer's subject. If you can fetch the Sha-256 hash, you could even use `Certificate`, with a tiny patch:
```diff
--- a/core/observables/certificate.py
+++ b/core/observables/certificate.py
@@ -35,6 +35,10 @@ its hash_sha256 value.
         """
         if hash_sha256 is None:
             hash_sha256 = hashlib.sha256(data).hexdigest()
+        if data is None:
+            return cls.get_or_create(
+                value="CERT:{}".format(hash_sha256),
+            )
         body = AttachedFile.from_content(
             BytesIO(data),
             "cert.der",
```

**Comment by threathive** [comment]
Interesting.. I was looking over some of that yesterday and it didn't seem like I could make this work. I will take another shot at it thanks !

**Comment by threathive** [comment]
depends on https://github.com/yeti-platform/yeti/pull/367

**Comment by tomchop** [comment]
Now that #367 has been merged, please rebase to see if Travis is happy again :)

File: plugins/analytics/public/circl_passive_ssl.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                                cert_sha1, results.settings)
+                            if cert_result:
+                                _info = cert_result.get('info', {})
+                                x509 = load_certificate(
+                                    FILETYPE_PEM, cert_result.get('pem'))
+                                hash_sha256 = ''.join(x509.digest(
+                                    'sha256').decode().lower().split(':'))
+
+                                cert_ob = Certificate.from_data(
+                                    data=None, hash_sha256=hash_sha256)
```
[the remark this answers is not in the document]
Comment (threathive) [reply plugins/analytics/public/circl_passive_ssl.py]: Did a update to the pr which includes this now.

File: plugins/analytics/public/circl_passive_ssl.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    for cert_sha1 in ip_details.get('certificates', []):
+                        try:
+                            cert_result = CirclPassiveSSLApi.fetch_cert(
+                                cert_sha1, results.settings)
+                            if cert_result:
+                                _info = cert_result.get('info', {})
+                                x509 = load_certificate(
+                                    FILETYPE_PEM, cert_result.get('pem'))
+
+                                DER = dump_certificate(FILETYPE_ASN1, x509)
```
Comment (p-l-) [inline plugins/analytics/public/circl_passive_ssl.py]: It is a good practice to use lowercase strings for variable names 
```suggestion
                                der = dump_certificate(FILETYPE_ASN1, x509)
```

File: plugins/analytics/public/circl_passive_ssl.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+                                cert_ob = Certificate.from_data(
+                                    data=DER)
```
Comment (p-l-) [inline plugins/analytics/public/circl_passive_ssl.py]:
```suggestion
                                    data=der)
```

File: plugins/analytics/public/circl_passive_ssl.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+                                subject = CertificateSubject.get_or_create(
+                                    value=_info.get('subject', ''))
+
+                                issuer = CertificateSubject.get_or_create(
+                                    value=_info.get('issuer', ''))
+
+                                links.update(observable.active_link_to(
+                                    cert_ob,
+                                    'cert_details',
```
Comment (p-l-) [inline plugins/analytics/public/circl_passive_ssl.py]: I'd suggest the use of the same names that are already used in IVRE plugin:
```suggestion
                                    'ssl-cert',
```

File: plugins/analytics/public/circl_passive_ssl.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                                    'circl_passive_ssl_query'))
+
+                                links.update(cert_ob.active_link_to(
+                                    subject,
+                                    'subject',
```
Comment (p-l-) [inline plugins/analytics/public/circl_passive_ssl.py]:
```suggestion
                                    'cert-subject',
```

File: plugins/analytics/public/circl_passive_ssl.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                                    'circl_passive_ssl_query'))
+
+                                links.update(cert_ob.active_link_to(
+                                    issuer,
+                                    'issuer',
```
Comment (p-l-) [inline plugins/analytics/public/circl_passive_ssl.py]:
```suggestion
                                    'cert-issuer',
```

**Comment by tomchop** [comment]
Thanks a lot @p-l- for taking a look at this!

**Pull request merged**